File: Remote/Source/Remote Elements/Commands/Command Subclasses/SwitchCommand.m
//
//  SwitchCommand.m
//  Remote
//
//  Created by Jason Cardwell on 3/25/13.
//  Copyright (c) 2013 Moondeer Studios. All rights reserved.
//
#import "Command_Private.h"
#import "RemoteController.h"
#import "Remote.h"
#import "RemoteElementExportSupportFunctions.h"
#import "RemoteElementImportSupportFunctions.h"

static int ddLogLevel   = LOG_LEVEL_DEBUG;
static int msLogContext = (LOG_CONTEXT_COMMAND | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);

@interface SwitchCommandOperation : CommandOperation @end

@implementation SwitchCommand

@dynamic target;

- (CommandOperation *)operation {
  return [SwitchCommandOperation operationForCommand:self];
}

- (SwitchCommandType)type {
  [self willAccessValueForKey:@"type"];
  NSNumber * type = [self primitiveValueForKey:@"type"];

  [self didAccessValueForKey:@"type"];

  return IntValue(type);
}

- (MSDictionary *)JSONDictionary {
  MSDictionary * dictionary = [super JSONDictionary];

  SetValueForKeyIfNotDefault(switchCommandTypeJSONValueForSwitchCommand(self), @"type", dictionary);
  SetValueForKeyIfNotDefault((self.type ?
                             self.target :
                             [[Remote existingObjectWithUUID:self.target
                                                     context:self.managedObjectContext] commentedUUID]),
                             @"target",
                             dictionary);

  [dictionary compact];
  [dictionary compress];

  return dictionary;
}

- (void)updateWithData:(NSDictionary *)data {
  /*
       {
           "class": "switch",
           "type": "remote",
           "target": "B0EA5B35-5CF6-40E9-B302-0F164D4A7ADD" // Home Screen
       }
   */

  [super updateWithData:data];

  [self setPrimitiveValue:@(switchCommandTypeFromImportKey(data[@"type"])) forKey:@"type"];

  NSString * target = data[@"target"];

  if (UUIDIsValid(target)) self.target = target;

}

@end

@implementation SwitchCommandOperation

- (void)main {
  @try {
    NSManagedObjectContext * moc        = _command.managedObjectContext;
    RemoteController       * controller = [RemoteController remoteController:moc];

    if (((SwitchCommand *)_command).type == SwitchModeCommand) {
      Remote   * remote = controller.currentRemote;
      NSString * mode   = ((SwitchCommand *)_command).target;

      remote.currentMode = mode;
      _success           = [remote.currentMode isEqualToString:mode];
    } else if (((SwitchCommand *)_command).type == SwitchRemoteCommand)   {
      Remote * remote = [Remote existingObjectWithUUID:((SwitchCommand *)_command).target context:moc];

      if (remote) {
        controller.currentRemote = remote;
        _success                 = controller.currentRemote == remote;
      } else _success = NO;
    } else
      _success = NO;

    [super main];
  } @catch(NSException * exception)   {
    MSLogError(@"command failed");
  }
}

@end

File: Remote/Source/Remote Elements/Commands/Command Subclasses/SwitchCommand.swift
//
//  SwitchCommand.swift
//  Remote
//
//  Created by Jason Cardwell on 3/2/15.
//  Copyright (c) 2015 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(SwitchCommand)
class SwitchCommand: Command {

  /** Holds the target's `uuid` when type is remote and the mode value when the type is mode */
  @NSManaged var target: String

  @objc enum SwitchType: Int { case Undefined = 0, Remote = 1, Mode = 2 }

  @NSManaged var primitiveType: NSNumber
  var type: SwitchType {
    get {
      willAccessValueForKey("type")
      let type = primitiveType
      didAccessValueForKey("type")
      return SwitchType(rawValue: type.integerValue) ?? .Undefined
    }
    set {
      willChangeValueForKey("type")
      primitiveType = newValue.rawValue
      didChangeValueForKey("type")
    }
  }

  /**
  JSONDictionary

  :returns: MSDictionary!
  */
  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()

    dictionary["class"] = "switch"
    setIfNotDefault("type", inDictionary: dictionary)
    switch type {
      case .Remote:
        if let targetRemote = Remote.existingObjectWithUUID(target, context: managedObjectContext!) {
          dictionary["target"] = target
        }
      case .Mode:
        dictionary["target"] = target
      default: break
    }

    dictionary.compact()
    dictionary.compress()

    return dictionary
  }

  /**
  updateWithData:

  :param: data [String:AnyObject]
  */
  override func updateWithData(data: [String:AnyObject]) {
    super.updateWithData(data)
    if let typeJSON = data["type"] as? String { self.type = SwitchType(JSONValue: typeJSON) }
    if let target = data["target"] as? String { self.target = target }
  }

  override var operation: CommandOperation {
    return SwitchCommandOperation(forCommand: self)
  }
}

extension SwitchCommand.SwitchType: JSONValueConvertible {
  var JSONValue: String {
    switch self {
      case .Undefined: return "undefined"
      case .Remote:    return "remote"
      case .Mode:      return "mode"
    }
  }

  init(JSONValue: String) {
    switch JSONValue {
      case SwitchCommand.SwitchType.Remote.JSONValue: self = .Remote
      case SwitchCommand.SwitchType.Mode.JSONValue:   self = .Mode
      default:                                        self = .Undefined
    }
  }
}